### ISSUE TITLE
Iterators and `async`/`await`

### ISSUE THREAD

**Issue by mentalisttraceur** [mentalisttraceur/python-with#1]
How could `with_` handle cases like this?

* ```python
  with manager as value:
      generator = action(value)
      yield from generator
  ```
* ```python
  async with manager as value:
      result = await action(value)
  ```
* ```python
  async with manager as value:
      async_generator = await action(value)
      yield from async_generator
  ```

**Comment by mentalisttraceur** [comment]
My initial reflex was to boil down the iterable variant of `with_` all the way down to this:

```python
def iterate_with(manager, iterator):
    with manager as value:  # hmm... wait a second...
        return (yield from iterator)  # ..can't use `value`!
    return None
```

That rather bothered me. I would guess that in the common case you might not need the context manager value, but seems like a pretty deal-breaking problem in the edge cases where it does matter.

So, okay, next improvement:

```python
def iterate_with(manager, action):
    with manager as value:
        return (yield from action(value))
    return None
```

Now `action` just needs to return an iterable, and has the option of using the context manager's `value`. So for example, `action` could be a generator, which is also nice because it is very consistent with regular `with_` - in both cases, we can basically just pull a `with` body into a function, give it one argument of the context manager's `as` value, and call it a day. Meanwhile, if we already have an iterator and don't need to use the context manager' value, the iterator can just be wrapped with `lambda _: iterator` to create a valid `action` for iterator-`with_`.

(For consistency with some idiomatic Python abbreviations, I might use the name `iwith` in actual releases instead of something like `iterate_with` or `iterator_with`.)

**Comment by mentalisttraceur** [comment]
The portability of the above `iterate_with` starts getting messy below Python 3.4, which added `yield from`.

Below 3.4, the `yield from` logic of properly forwarding `.send` and `.throw` calls has to be implemented manually. But now that I've written [`yield-from-as-an-iterator`](/mentalisttraceur/python-yieldfrom), this is handled down to about 2.6:

```python
from yieldfrom import yield_from as _yield_from

def iterate_with(manager, action):
    result = None
    with manager as value:
        yield_from = _yield_from(action(value))
        for value, handle_send, handle_throw in yield_from:
            sent = None
            try:
                sent = yield value
            except:
                if not handle_throw(*_exc_info()):
                    raise
            handle_send(sent)
        result = yield_from.result
    if result is not None:
        raise StopIteration(result)
```

It's also handled on 2.5 as far as I'm concerned (someone would need to polyfill the builtin function `next` for `yield_from`, which I think is trivial enough to implement that it's the job of whoever is still running 2.5 in this day and age).

Below 2.5 things get painful, because at that point `yield` stops being bidirectional.

**Comment by mentalisttraceur** [comment]
Actually, it seems like `iterate_with` should try to ensure that it always closes the iterator it gets! **[Edit: two comments later I realized that this is wrong.]** This handles cases such as the `action` being a generator function that itself has a context manager internally:

```python
def iterate_with(manager, action):
    with manager as value:
        with _ClosingIfClosable(action, value) as iterator:
            return (yield from iterator)
    return None

class _ClosingIfClosable:
    __slots__ = ('_action', '_value', '_iterator')

    def __init__(self, action, value):
        self._action = action  
        self._value = value   

    def __enter__(self):
        self._iterator = iter(self._action(self._value))
        return self._iterator

    def __exit__(self, exception_type, exception, traceback):
        try:
            close = self._iterator.close
        except AttributeError:
            return False
        close()
        return False
```

[PEP-533](https://peps.python.org/pep-0533/) has some good coverage of this problem (needing to defensively wrap every `for` that might be given a generator to iterate on in a `with` that can close the iterator if it is closable).

For me this adds evidence in favor of "yes" to the "should I provide an iterator variant of `with_`?" question, because this is a nuance that can cause subtle bugs if not handled but is obscure enough that many people wouldn't think of it in many situations (moreover, if you look at the edit history of this comment, it took me several tries to correctly protect against all edge cases).

So this is valuable *even on current and fully featured Python implementations*.

**Comment by mentalisttraceur** [comment]
It's tempting to think that `iterate_with` doesn't need to wrap its iteration with any closing-if-closable protections, because it is so simple that it should be effectively transparent to `.close`. **[Edit: in the next comment I realize that `iterate_with` doesn't need to automatically do close-if-closable because `iterate_with` itself can be composed to wrap the underlying iterator with a closing context manager.]**

(I tried to convince myself of this, because it would be *really* nice if I didn't have to take care of backporting the `with _ClosingIfClosable` *inside of the backport of `with`*, where I cannot rely on `with` itself...  and experience shows that Python reimplementations sometimes struggle with getting `else` and `finally` clauses or `try` fully correct on first try, so really I've only got `try` and `except`.)

But it turns out that this really is a problem (even ignoring the backports and incomplete Pythons for now, just thinking about the situation on the latest Python versions with robust native `with` and `yield from`). Consider this example:

```python
class ClosableExample:
    def __init__(self, unused):
        del unused
    def __iter__(self):
        return self
    def __next__(self):
        raise Exception('oops an error')
    def close(self):
        print(f'closing! {self!r}')

class ContextManagerExample:
    def __enter__(self):
        print(f'entering! {self!r}')
    def __exit__(self, exception_type, exception, traceback):
        print(f'exiting! {self!r}')

def iterate_with_1(manager, action):
    with manager as value:
        return (yield from action(value))
    return None
```

Now this fails (notice that "closing!" is never printed, neither when the error happens and nor even when manually closing afterwards):

```python
>>> i = iterate_with_1(ContextManagerExample(), ClosableExample)
>>> next(i)
entering! <__main__.ContextManagerExample object at 0x7f595c59b9d0>
exiting! <__main__.ContextManagerExample object at 0x7f595c59b9d0>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in iterate_with_1
  File "<stdin>", line 7, in __next__
Exception: oops an error
>>> i.close()
>>> 
```

This means that we can't actually impose the right protections from the outside. That close-if-closable really does have to be done *inside* the `iterate_with`:

```python
def iterate_with_2(manager, action):
    with manager as value:
        with _ClosingIfClosable(action, value) as iterator:
            return (yield from iterator)
    return None
```
```python
>>> i = iterate_with_2(ContextManagerExample(), ClosableExample); next(i)
entering! <__main__.ContextManagerExample object at 0x7f595c59b3a0>
closing! <__main__.ClosableExample object at 0x7f595c59b670>
exiting! <__main__.ContextManagerExample object at 0x7f595c59b3a0>
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in iterate_with_2
  File "<stdin>", line 7, in __next__
Exception: oops an error
```

(Of course, one downside of this is that in the non-error case, `.close` ends up being called twice. But `.close` is supposed to be basically idempotent so that should be fine.)

(In CPython this is deceptively "fixed" by just adding a `__del__ = close` to my `class ClosableExample` definition, which makes my example more like a normal generator, but the whole reason why an automatic "closing" is nice is that it creates deterministic cleanup that *doesn't rely on garbage collection*. You can add the `__del__` to this class and test this on PyPy to still see the same problem.)

**Comment by mentalisttraceur** [comment]
Another downside of adding automatic closing is that composition no longer works the same way, and it's no longer nearly as straightforward of a translation from `with` (statement+block) to `iterate_with` (function+closure).

I mean that right now, there's this edge case in Python where you can do this:

```python
with manager:
    for thing in iterator:
        if thing == 'some magic value':
            break
        ... # Do stuff with values up to "some magic value"

# iterator is still usable
for thing in iterator:
    ...  # Do stuff with values after "some magic value"
```

I don't think it comes up very often, but I know I've written a parser before that kept passing the same iterator (of token streams) up and down the call tree for parsing functions to consume the head of. And it's not a big stretch to imagine someone wanting to compose a context manager with that somehow.

So with `iterate_with`, you can put that first `with` body into a generator function, let's say called `consume_first_part`, but if there's automatic closing, you can't just write `iterate_with(manager, lambda _: consume_first_part(iterator)` to replace the first `with` block, because that would change behavior.

Of course the workaround is to just define a wrapper iterator that ignores `.close`: `iterate_with(manager, lambda _: do_not_auto_close(consume_first_part(iterator)))`. Of course at a minimum, that wrapper has to do full passthrough of `__next__`, `send`, and `throw` methods to not lose or break functionality.)

The other direction is: given an `iterate_with` *without* automatic closing, can we somehow inject `_ClosingIfClosable` when needed? Well, turns out the answer is yes:

```python
iterate_with(manager, lambda value: iterate_with(_ClosingIfClosable(action, value)), lambda iterator: iterator))
```

Of course, if you don't care about the two benefits of something like my `_ClosingIfClosable` (the big one is the "if closable" part so you can apply it without having to know if you were given a closable, but the small one is that it will delay the `action(value)` call until the `__enter__` method, and thus close a tiny race condition in case `action(value)` constructs an object which already needs `close` for cleanup before the first call to its `__next__`), then you can just use `contextlib.closing` from the standard library in the same way: 

```python
iterate_with(manager, lambda value: iterate_with(contextlib.closing(action(value)), lambda iterator: iterator))
```

Admittedly, this is... *painful* boilerplate, I'd say on balance this is worse boilerplate than the boilerplate of a `do_not_auto_close` iterator wrapper (at least in terms of being able to decode and follow it if you don't already know), but at a quick glance (read: some basic manual testing) it seems robust.

But it is great, far better than the automatically-close-by-default behavior, for implementation simplicity. Because if I had to make my fundamental building block  a *double* `with` context, then the backporting/backpiling to older and incomplete Pythons was going to get way more complicated and difficult - but now I am free to scope that problem back down to a single `with`.

There's also something nice about the portable function-oriented thing I'm doing here more closely mirroring the semantics and behavior of native Python `with` - it retains a straightforward translation of `with` statements to `iterate_with` calls.

**Comment by mentalisttraceur** [comment]
So, if we wanted to implement `iterate_with` below 2.5 (or more precisely, on any Pythons without a working bidirectional `yield`... of the ones that are "living" and seeing real production use that might just be Skulpt), we have a couple options.

1. Reconsider and decide to *not* do that instead. This is the best option for most intents and purposes.

2. Spend two weeks reimplementing all generator logic from scratch, as an object that follows the iterator protocol and the [PEP-342](https://peps.python.org/pep-0342/) protocol of bidirectional generators.

    This is almost never the best option. Naturally I did it anyway - expect to see that in the release. My only defense is that it's a learning exercise that opens some actually valuable doors - for example, I am now much more equipped to write a transpiler that turns generators into *pickleable* objects that behave like generators, or to write a library that provides bidirectionally transparent `map`/`filter`/`functools.reduce`.

---

I then spent several days figuring out how to adapt these various backports to the various other Pythons that see production use: as far as I know that's MicroPython, Brython, Skulpt, and Transcrypt. But that has grown into a sufficiently open-ended combinatoric explosion that maybe if finally broke my will to support these incomplete Pythons:

1. Transcrypt ends up needing special-casing. Its own transpilation of `with` has at least two errors (`with foo as value` does `value = foo` instead of `value = foo.__enter__()` and control flow like `return` inside the `with` will just bypass context manager exit). Also, if you use native JavaScript generators your `iterate_with` will not fulfill the same Python interfaces - that's probably fine... but it's yet another decision that needs to be made (and is mentally expensive to pick one decision for the general case). In the end I basically ended up deciding that I might as well write the JavaScript by hand, because it ends up being cleaner, clearer, and less needlessly quirky. I already wrote that JS so maybe it'll make it into the main branch and sdist at some point.

2. MicroPython seems to have a fine `with` and `yield from` natively nowadays - there are some quirks/deviations from how CPython works in some error edge cases, but it's good enough that I don't consider it necessary to do anything here. Great work on their part.

3. Brython... I always had great impressions of Brython, and they're still mostly great, but I've already opened two bug reports for them in the last couple days as I've been testing their native `yield from` behavior (they fixed one within 24 hours, to their credit). The backport of `iterate_with` to their current variant isn't hard though, it's just the 2.6-3.2 implementation except we can (and must) use `return stop.value` instead of re-raising `StopIteration` (this is actually nicer, and is how I initially wrote the backport once I realized that `yield-from-as-an-iterator` actually wasn't quite fit for the purpose because of how the `close` and `throw` handling actually had different `StopIteration` handling and so couldn't be wrapped in the same exception exit path - it's just that in regular Python, `return` with a value was not allowed in a generator and `StopIteration` was allowed to get raised by a generator until `yield from` was added, while Brython technically has a `yield from` and thus follows the 3.3+ rules of allowing `return` with a value but turning `StopIteration` into `RuntimeError`).

4. Skulpt. Skulpt skulpt skulpt. Nothing quite breaks my motivation to support incomplete Pythons like Skulpt does. Originally, Skulpt was one of the reasons why I wrote `with-as-a-function` - it didn't have `with`. Nowadays it has `with`, and it might even work correctly, maybe even all the time - I don't know if they've fixed issues such as "`__exit__` gets called twice sometimes". They don't have bidirectional generators, so the backport has to be the 2.2-2.4 `iterate_with` - the hand-written class which mimicks a generator, but forward-ported to use Python 2.6+ syntax, and then hand-tweaked to manage various Skulpt quirks like the built-in `next` discarding the arguments of `StopIteration`.

**Comment by mentalisttraceur** [comment]
Now, since my initial sequence of posts about two weeks ago, I found many frustrating edge cases in Python's `yield from`, and how it interacts with `with`. Here's the "specification" from PEP-380, adapted to show what CPython actually does:

```python
try:
    value = next(iterator)
except StopIteration as stop:
    result = stop.value
else:
    try:
        while True:
            try:
                sent = yield value
            except GeneratorExit:
                try:
                    close = iterator.close
                except AttributeError:
                    pass
                except:
                    # Python swallows these exceptions and sends them
                    # to `sys.unraisablehook`. The default hook prints it
                    # to standard error. Kinda like this:
                    print(f"Exception ignored in {iterator!r}", file=sys.stderr)
                    traceback.print_exc()
                    # The original `yield from` patch predates `sys.unraisablehook`
                    # but you can see it used `PyErr_WriteUnraisable` back then
                    # ( https://bugs.python.org/file24214/f8349cbc1b26.diff ).
                    # PyPy emulates this, but other Pythons like MicroPython
                    # miss this corner case. Personally I think this behavior is
                    # bad design and Python was wrong to do it this way.
                else:
                    try:
                        close()
                    except BaseException as exception:
                        # Unlike the example code in PEP-380, Python's actual
                        # C implementation does not set the `GeneratorExit`
                        # instance as the context of errors from the close.
                        exception.__context__ = None
                        raise  # bubble up exception from close()
                raise  # bubble up the GeneratorExit
            except:
                try:
                    throw = iterator.throw
                except AttributeError:
                    # If we re-raised the thrown-in exception in here, we'd
                    # need to clear this AttributeError from `.__context__`
                    # to match actual Python behavior.
                    throw = None
                except:
                    # In Python, this generator body actually never sees
                    # exceptions from the .throw lookup on the iterator:  
                    # those just get raised directly from the outer .throw().
                    # I'm unsure if this was on purpose or just a side-effect
                    # of how they optimized delegation at the C level.
                    # Other Pythons like PyPy and MicroPython end up   
                    # treating these other errors from accessing .throw
                    # the same way as errors from calling throw.
                if throw is None:
                    raise  # bubble up the thrown-in exception
                # Deferring the `throw` call until we exit this `except` once
                # again saves us from a situation where we would have
                # to clear  `__context__` to match native `yield from`.  
                thrown = _exc_info()
            try:
                if thrown is not None:
                    value = throw(*thrown)
                elif sent is not None:
                    value = iterator.send(sent)
                else:
                    value = next(iterator)
            except StopIteration as stop:
                result = stop.value
                break
    except StopIteration as stop:
        # This RuntimeError doesn't actually get raised in the generator body,
        # but as I understand it, it happens as soon as the delegated-to
        # generator raises a StopIteration and propagates out without the
        # delegating generator getting a chance to catch it.
        raise RuntimeError("generator raised StopIteration") from stop
```

My "favorite" is that any error when looking up the `.close` attribute on the iterator (*besides* `AttributeError`) is just suppressed and printed as unraisable on standard error.

I think this is awful design because:

1. If there is an unexpected error in *getting* `.close`, it may very well be that there was supposed to be cleanup that now couldn't be done, which is *not* "there was no closing to do, all good, move on" - it is "closing failed!"
2. Inconsistently, errors from *calling* `.close` *don't* get suppressed, which is obviously sensible because such errors mean "closing failed!" and the closing-trying code might want to know about it.
3. For most code, this is an entirely silent swallowing - for code that does not go out of its way to intercept that "unraisable" error logging, that error might as well not exist.

But there are other quirks - for example, in CPython, looking up `.throw` on the iterator fails (again besides `AttributeError`), this failure happens "outside" the generator that is doing the `yield from`, so our `iterate_with` stays "open" and the context manager never sees the error, as if it just never got thrown - but in almost all other Pythons, presumably because they actually try to follow the specification in PEP-380 as if it's the real specification and not just an afterthought for communicating the original C implementation, such attribute lookup errors are bubbled up in the generator, so in our `iterate_with` the context manager sees the error and if it doesn't suppress it then the generator closes.

A "fun" special case of that last problem is: what if looking up (not calling!) `.throw` raises `StopIteration` for some reason? Well in Python, that will be *externally entirely ambiguous with* `.throw` succeeding and resulting in a clean finishing of the generator, even though in reality it was never thrown into the generator. Contrast PyPy, this will get treated as a clean exit from the generator (including exiting the context manager as if no exception is bubbling up), presumably because the PyPy devs did the equivalent of `iterator.throw(...)` as one operation with the same exception handling, as if that's the same as separating the attribute lookup from the attribute call.

What an utter mess.

Almost no other Python gets all of these edge cases the same.

**Comment by mentalisttraceur** [comment]
So the conclusion I reached after

1. meticulously testing across multiple versions and implementations of Python to discover all this behavior,
2. detouring into the CPython source (once the suppressed "unraisable" error complication made it overwhelmingly clear that the PEP-380 specification wasn't even close to the real deal and that I needed to see first-hand what other possibilities too unreasonable for me to even conceive of were lurking in there), and
3. struggling to faithfully replicate all of it until the "unraisable" [censored] [censored] finally made something snap inside me,

since there's no way to use native `yield from` without these deviations in the error corner cases anyway, the portable backport doesn't need to replicate any specific one of those behaviors, so I am not going to.

Exceptions *from* the iterator (flying out of calls to the `__next__`, `send`, `throw`, or `close` methods) will be propagated or handled correctly and consistently in all implementations. But if the iterator implements the relevant interfaces incorrectly - that is, if just *accessing* `.throw`, or `.close` raises an exception other than `AttributeError`, I will not and cannot commit to exact bug-for-bug, all-observable-behavior matching of any given implementation. In those error (or even meta error) corner cases, you will get native `yield from` behavior of your Python if it has a working native `yield from`, otherwise you will get a sensible backport whose behavior in those corner cases is at least as sensible as existing mainstream `yield from` implementations.

**Comment by mentalisttraceur** [comment]
Turning my attention back to the `async`/`await` question, I think I'm basically putting that aside for now. But some thoughts:

0. `awith` and `aiwith` would be the names - as much as I dislike abbreviation, this follows very strongly established Python naming idioms.

1. I think it's fine for a "minimum portable foundation" to basically insist that all of the inputs are async: an async context manager, an action that returns an awaitable, and in the iteration case, an async iterator/generator. I think it's fine to have just one interface like this because it is always possible and even simple to wrap or proxy a synchronous object to fulfill the interfaces demanded by `async`/`await` syntax, because a synchronous function or method can just be called by an asynchronous function or method. This is boilerplate, but it is boilerplate that can be solved by a separate library for all cases where higher-order functions need to plug such things in.

2. However, a usability/developer experience improvement would be to detect either all cases or the unambiguous cases. Certainly we could check if the manager's type has `__aenter__` and `__aexit__` and use `async with`, otherwise use `with`; ditto for checking if the iterable's type has `__aiter__` or maybe trying to call `aiter` on the iterable. For `action`, the actual most-robust check is "is the returned value awaitable or not?" Technically, in all of those cases, we can just go with the result of that check, and in the rare case where a user has an object that fulfills the async interface but wants it to be treated as the sync case, that object can also be wrapped to "hide" the async interface. Ultimately it's a question of a) how much complexity does it add to check for the cases and b) how frequently would users be benefited or frustrated with either choice. I think the latter is the more important consideration and that the answer is that "I have a sync-only thing, I want to shove it in here and have it treated as sync" is much more likely than "I have an either-sync-or-async thing, and I want it treated as sync".

3. But if we go with the second option, then what should `awith` or `aiwith` do if each object is sync? Sync context manager and sync action (and sync iterable) - should `awith` still require an `await` and should `aiwith` still require `async for`? I lean strongly towards yes. I think the answer is "no" for now, but I suspect there *is* a use-case for that, same as I eventually found a use-case for it with `compose` and `acompose` - once people have the power of higher-order functions, eventually they want *ergonomics* around those functions, quick helper functions/objects/syntax/sugar, and the implementation of *that* usually benefits from having something that is dynamically either async or not based on whether you gave it any async things or only sync things. Once that becomes apparent for `with-as-a-function`, I can perhaps follow the pattern I did in `compose` - `sawith` and `saiwith`.

4. `async` generators in Python are still not allowed to do `yield from`, so it's currently actually non-trivial to write a higher-order function that wraps an async generator with a context manager while preserving. bidirectional transparency. Amusingly, the implementation I've written for backporting generator and yield-from behavior probably could be easily adapted to emulate `yield from` on asynchronous generators.

5. I will not bother "backporting" `async` stuff below 3.5 or to Pythons without native `async`/`await` any time in the foreseeable future. This is a rule I set for myself in all of my packages that use `async`/`await`, because... well, you see how I get obsessed with porting code. It's unhealthy and I have to have lines somewhere. What's neat is that bidirectional generators are sufficient foundation to implement asynchronous coroutines - in Python this is actually how the implementation works, and `async` functions are actually basically just generators under the covers, and `await` is basically just `yield from`. So given a proper backport of generators and `yield from`, which I've now exhaustively done, we can implement a backpiler which transpiles `async`/`await` code into generators, maybe slaps some decorators on them like `asyncio` did in 3.4, and then you just need an event loop. And for that kind of backport, you can use `iwith` to help you wrap of each of those backpiled generators with context managers (in other words, for targets without native `async`/`await`, `iwith` *is* `awith`, and `aiwith` is probably just an `iwith` iterating on a wrapper iterator that uses `iwith` in all of the async generator protocol methods, although I'd have to think about that more rigorously). Point being, if it matters enough, someone can write a backpiler which can benefit a lot from just `iwith`.

**Comment by mentalisttraceur** [comment]
So, verdicts:

1. `iterate_with`, named `iwith` to match some Python conventions, is already implemented, and will be released soon as v1.1.0.

2. `awith` and `aiwith` will come some day maybe, sooner and more likely if anyone asks.

    P.S. I do think the need for `awith` and `aiwith` is lower - the need was stronger for `with_` and `iwith` because there are Pythons *today* where I cannot get reliably correct *`with`* behavior, and the only way to fix that is a *non-syntax* interface which I can port to those all Pythons we might use. But Pythons that implement `async`/`await` correctly often also implement `with` correctly. (But this isn't universally true: for example, Transcrypt transpiles `async`/`await` directly to the same JavaScript keywords, but it still doesn't transpile `with` correctly.)
